Source/NSPasteboardItem.m

```diff
@@ -0,0 +1,177 @@
+   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+/*
+ * FIXME: this class is useless until NSPasteboard implements
+ * NSPasteboardReading + NSPasteboardWriting.
+ */
+
+#import <AppKit/NSPasteboardItem.h>
+#import <Foundation/Foundation.h>
+
+@implementation NSPasteboardItem
+- (id)init
+{
+  if ((self = [super init]))
+    {
+      _providerMap = [[NSMutableDictionary alloc] init];
+      _dataMap = [[NSMutableDictionary alloc] init];
+      _types = [[NSMutableArray alloc] init];
+    }
+
+  return self;
+}
+
+- (NSArray *)types
+{
+  return [[NSArray arrayWithArray: _types] copy]; 
+}
+
+- (BOOL)setDataProvider:(id<NSPasteboardItemDataProvider>)dataProvider
+               forTypes:(NSArray *)types
+{
+  if (![dataProvider conformsToProtocol: @protocol(NSPasteboardItemDataProvider)])
+    {
+      NSLog(@"Pasteboard item data provider %@ must conform to"
+             "NSPasteboardItemDataProviderProtocol", dataProvider);
+      return NO;
+    }
+
+  for (NSUInteger i = 0; i < [types count]; i++)
+    {
+      NSString *type = [types objectAtIndex: i];
+      [_providerMap setObject: dataProvider forKey: type];
+    }
+
+  [_types addObjectsFromArray: types];
+  return YES;
+}
+
+- (NSString *)availableTypeFromArray:(NSArray *)types
+{
+  // FIXME: this would require a dependency on UTIs
+  return nil;
+}
+
+- (BOOL)setData:(NSData *)data forType:(NSString *)type
+{
+  if (![data isKindOfClass: [NSData class]])
+    return NO;
+  
+  [_dataMap setObject: data forKey: type];
+  return YES;
+}
+
+- (BOOL)setString:(NSString *)string forType:(NSString *)type
+{
+  if (![string isKindOfClass: [NSString class]])
+    return NO;
+  
+  [_dataMap setObject: string forKey: type];
+  return YES;
+}
+
+- (BOOL)setPropertyList:(id)propertyList forType:(NSString *)type
+{
+  if (![NSPropertyListSerialization propertyList: propertyList
+                                isValidForFormat: NSPropertyListXMLFormat_v1_0])
+    return NO;
+
+  [_dataMap setObject: propertyList forKey: type];
+  return YES;
+}
+
+- (NSData *)dataForType:(NSString *)type
+{
+  id object = [_dataMap objectForKey: type];
+  if ([object isKindOfClass: [NSData class]])
+    return (NSData *)object;
+  
+  return [NSPropertyListSerialization dataWithPropertyList: object
+                                          format: NSPropertyListXMLFormat_v1_0
+                                         options: 0
+                                           error: NULL];
+}
+
+- (NSString *)stringForType:(NSString *)type
+{
+  id object = [_dataMap objectForKey: type];
+  if ([object isKindOfClass: [NSString class]])
+    return (NSString *)object;
+  
+  NSData *data = [self dataForType: type];
+  return [[[NSString alloc] initWithData: data
+                                encoding: NSUTF8StringEncoding] autorelease];
+}
+
+- (id)propertyListForType:(NSString *)type
+{
+  id object = [_dataMap objectForKey: type];
+  if (![object isKindOfClass: [NSData class]])
```

Comment:
Not sure whether this is correct. In most other places NSString and a few other types would be valid here too.